contributors/devel/sig-instrumentation/migration-to-structured-logging.md

```diff
@@ -0,0 +1,346 @@
+```go
+package klog
+
+// InfoS structured logs to the INFO log.
+// The msg argument used to add constant description to the log line.
+// The key/value pairs would be join by "=" ; a newline is always appended.
+//
+// Examples:
+// >> klog.InfoS("Pod status updated", "pod", klog.KObj(pod), "status", "ready")
+// output:
+// >> I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kube-system/kubedns" status="ready"
+func InfoS(msg string, keysAndValues ...interface{})
+
+// ErrorS structured logs to the ERROR, WARNING, and INFO logs.
+// the err argument used as "err" field of log line.
+// The msg argument used to add constant description to the log line.
+// The key/value pairs would be join by "=" ; a newline is always appended.
+//
+// Examples:
+// >> klog.ErrorS(err, "Failed to update pod status")
+// output:
+// >> E1025 00:15:15.525108       1 controller_utils.go:114] "Failed to update pod status" err="timeout"
+func ErrorS(err error, msg string, keysAndValues ...interface{})
+
+// KObj is used to create ObjectRef when logging information about Kubernetes objects
+// Examples:
+// >> klog.InfoS("Pod status updated", "pod", klog.KObj(pod), "status", "ready")
+// output:
+// >> I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kube-system/kubedns" status="ready"
+func KObj(obj KMetadata) ObjectRef
+
+// KRef is used to create ObjectRef when logging information about Kubernetes objects without access to metav1.Object
+// Examples:
+// >> klog.InfoS("Pod status updated", "pod", klog.KRef(podNamespace, podName), "status", "ready")
+// output:
+// >> I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kube-system/kubedns" status="ready"
+func KRef(namespace, name string) ObjectRef
+
+// ObjectRef represents a reference to a kubernetes object used for logging purpose
+// In text logs it is serialized into "{namespace}/{name}" or "{name}" if namespace is empty
+type ObjectRef struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// KMetadata is a subset of the kubernetes k8s.io/apimachinery/pkg/apis/meta/v1.Object interface
+// this interface may expand in the future, but will always be a subset of the
+// kubernetes k8s.io/apimachinery/pkg/apis/meta/v1.Object interface
+type KMetadata interface {
+	GetName() string
+	GetNamespace() string
+}
+```
+
+## Migration
+
+1. Change log functions to structured equivalent
+1. Remove string formatting from log message
+1. Name arguments
+1. Use `klog.KObj` and `klog.KRef` for Kubernetes object references
+1. Verify log output
+
+## Change log functions to structured equivalent
+
+Structured logging functions follow a different logging interface design than other functions in `klog`. They follow
+minimal design from [logr] thus there is no one-to-one mapping.
+
+Simplified mapping between functions:
+* `klog.Infof`, `klog.Info`, `klog.Infoln`, `klog.InfoDepth` -> `klog.InfoS`
```

Comment:
The "depth" ones are challenging because they presume hat popping up the call-stack is possible.  I count about 8 such files that make these calls in k/k.  Do you link we need an `InfoS` variant that includes depth?  That would be ugly to add to logr because it tries to abstract multiple implementations, but maybe that's something we should tackle in logr and klog, if it is a missing capability...

Reply:
> Several of those are semi-global logging wrappers (1, 2, 5) that are better off just having an implementation that can do caller-skip (e.g. zap). I'm not sure we want the caller-skip at the interface level, though.

These ones seem like valid use-cases to me.  In fact, I wrote the first version of (5), I think.  The ability to strip frames from the callstack seems pretty necessary for any sort of logging util helper or adapter.

I don't quite grok your comment.  How would an implementation that can skip callframes be utilized under something like `logr` without a function for it in the interface?

In other words, should logr and klog grow new methods for `InfoDepth(depth, msg, string...)` and `ErrorDepth(depth, err, msg, string...)` ?

Reply:
Hmm... let me rephrase a bit.  I think I misinterpreted somethings, and miscommunicated others.  I've rewritten this comment a few times, so it's possible there's something I'm not quite articulating correctly.

I agree that they're valid use cases, but:

All the mentioned numbers follow a pattern of "here's a set of helper functions where all the functions have a consistent caller skip".  The caller skip isn't varying per-function in the file.  I was originally (above) suggesting something like

```go
var logWrapper = klog.WithSkip(4)

func Info(msg) {
  logWrapper.InfoS(prefix+msg)
}
```

However, that only works when we're not redirecting to logr.  That's not too big a deal -- it preserves the status quo, punts on putting things into our interface, and should be doable with the way the internals of klog work.

We could consider adding the `WithSkip` to logr, which makes the caller skip compose and feels a lot more like the rest of logr's interface.  However, this is a bit annoying to implement in some logging backends (see below the fold), and still feels a bit brittle (again, below the fold).

<details>

<summary> [below the fold] Why I'm kinda against adding the extra Depth methods </summary>

My objections/uneasiness with this has two parts:

The first is that adding depth functions feels a lot like leaking implementation details through to the user.  This kinda paints us into the corner in terms of always explicitly supporting caller skip like this -- I could see a world where we decide to drop caller in favor of logger names + stack traces for errors or a world where we decide to do "capture caller into context".

The second is that that per-function caller skip is an easy interface to accidentally misuse:

- It's easy to miss/forget the `Depth` in one of your functions.

- It doesn't compose (you can't wrap helpers in helpers and still get useful information, which is something I've seen with similar types of functionality in codebases like controller-runtime).

- It's sufficiently brittle that I'd actively encourage folks not to use it -- it relies on exact code structure not changing, which isn't something we can enforce with the compiler, lint for, etc.  In the case of the etcd logger, if the internals of the etcd client change, we suddenly lose the usefulness of our logging information.  That's not really an interface I want people writing against.  It's also wildly unclear where the constant `4` comes from.

Moving on to practical matters: how much we care about people actually being able to implement this, or would we rather they just ignore it?  Zap doesn't support per-function caller skip OOTB, but it's doable with some hacking (you'd call `entry := logger.Check(...); entry.Caller = /* calculate caller yourself */`, or maybe could finagle a custom `zapcore.Core` to do it), logrus doesn't even *do* caller skip AFAICT, etc.  Do we want to continue to support delegating to logr?  It's not clear how much of a goal that it from the KEP (I think we should, or at least have some similar functionality).

</details>

<details>

<summary>I think we can do better here if/when we have Context</summary>

A similar idea whence to draw inspiration is of Rust's [track_caller](https://github.com/rust-lang/rfcs/blob/master/text/2091-inline-semantic.md#propagation-of-tracker).  While the original motivation wasn't quite the same as with Go, it has nice ergonomics: you stick a "#[track_caller]" on a function, and it gets propagated downward to any functions that also have `#[track_caller]`.  It composes nicely (since the caller tracking propagates downwards) and doesn't require explicit numbering for caller skip (see below the fold).

If we draw inspiration from the link above, we get an option that isn't quite doable with today's interface, but falls out naturally from the key words "propagates downwards": context.

```go
func MyHelper(ctx context.Context, msg string) {
    // save the caller if there's not one already, otherwise this becomes a no-op
    ctx = log.TrackCaller(ctx)
    klog.InfoSC(ctx, msg)
}

// also have a
// - log.ResetCaller, which resets the propogation.
// - log.BrittleTrackCallerWithSkip, which does what it says on the tin
```

Obvs this doesn't compose if the function can't take context, and requires context on the logger (a change to the interface).  However, it does mean that individual logging impls don't need to know how to grab caller information -- they just pull it out of the context or whatever.

Side benefit is that this also fixes the composability of things like `util/naming/from_stack.go`, which currently break when wrapped in helpers.

</details>

Reply:
> We could consider adding the WithSkip to logr                                                                                                                 

The problem I have with this is that it doesn't feel like a property ofthe
logger object, but of the call-site.  I guess if you are wrapping a Logger in
some other type, you can argue that this IS part of the logger (or as you
characterized - it doesn't change).  If there's a one-off logging helper
function, you can always make a temporary object (though that can be
allocation-heav if it is done a lot).  E.g.

```
func logHelper(log logr.Logger, foo Foo) {
  logr.SkipFrames(1).Info("foo message", "address", foo.Addr(), "phone", foo.Phone)
}
```

> per-function caller skip is an easy interface to accidentally misuse:

OK, you convinced me.

> how much we care about people actually being able to implement this, or would we rather they just ignore it?

I suspect some log libs will handle it and some will no-op it.  I think that's
fine, honestly.  There will be pressure or there won't, but at least it is
_possible_ to DTRT

Reply:
> it can be alloc-heavy if it is done a lot

I would *hope* escape analysis and friends would prevent a heap allocation here.